src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -245,72 +245,48 @@ void addArg(RefKind refKind, BoundExpression expression)
                             return _factory.AssignmentExpression(output, _factory.Call(input, e.Property.GetMethod, _factory.Literal(e.Index)));
                         }
 
-                    case BoundDagIndexerEvaluation { IndexerAccess: null, IndexerSymbol: null } e:
-                        {
-                            // array[int]
-                            Debug.Assert(input.Type?.IsSZArray() == true);
-                            BoundExpression access = _factory.ArrayAccess(input, makeImplicitIndexArgument(e.Index, e.LengthTemp));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
-
-                    case BoundDagIndexerEvaluation { IndexerSymbol: PropertySymbol indexer } e:
-                        {
-                            // this[int]
-                            BoundExpression access = _factory.Indexer(input, indexer, makeImplicitIndexArgument(e.Index, e.LengthTemp));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
-
-                    case BoundDagIndexerEvaluation { IndexerAccess: BoundIndexerAccess indexerAccess } e:
+                    case BoundDagIndexerEvaluation e:
                         {
                             // this[Index]
-                            BoundExpression access = makeIndexerAccess(e.Syntax, indexerAccess, makeExplicitIndexArgument(e.Index, e.Index < 0));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
+                            // this[int]
+                            // array[Index]
+                            _localRewriter.AddPlaceholderReplacement(e.ReceiverPlaceholder, input);
+                            _localRewriter.AddPlaceholderReplacement<BoundListPatternUnloweredIndexPlaceholder>(e.ArgumentPlaceholder, makeUnloweredIndexArgument(e.Index));
```

Comment:
>makeUnloweredIndexArgument(e.Index) [](http://example.com/codeflow?start=135&length=35)

I think we can simply update the node with this index value as we do that for length access and avoid all that complexity around "unlowered" placeholders.